### PR TITLE
mon,mgr: make mgr controller adjust both pg_num and pgp_num in a consistent way

### PR DESCRIPTION
**Pull request by liewegas**
- pgp_num tracks pg_num unless/until it is set explicitly different than pg_num
- mgr adjusts pg_num upward aggressively (splits are cheap) (no change here)
- mgr adjusts pg_num downward by 1 pg at a time (merges are done carefully) (no change here)
- mgr adjusts pgp_num up or down, throttling based on the new global target_max_misplaced_ratio option
- mgr/balancer/max_misplaced is now replaced by target_max_misplaced_ratio

### CODE REVIEW

File: src/mgr/DaemonServer.cc [hunk 1]
```diff
@@ -2261,14 +2261,13 @@ void DaemonServer::adjust_pgs()
...
+			 p.get_pg_num_target() < p.get_pg_num()) {
 		dout(10) << "pool " << i.first
 			 << " target " << p.get_pg_num_target()
 			 << " pg_num " << p.get_pg_num()
-			 << " - pg_num_pending != pg_num, waiting"
+			 << " - decrease and pg_num_pending != pg_num, waiting"
 			 << dendl;
-		// FIXME: we might consider allowing pg_num increases without
-		// waiting for the previously planned merge to complete.
 	      } else if (p.get_pg_num_target() < p.get_pg_num()) {
```
Comment (xiexingguo) [inline src/mgr/DaemonServer.cc]: I guess the check for permitting a pg merge would then need to be changed to:

```c++
else if (p.get_pg_num() == p.get_pg_num_pending() &&
           p.get_pg_num_target() < p.get_pg_num())
```

File: PendingReleaseNotes [hunk 2]
```diff
@@ -72,9 +72,12 @@
 
     ceph auth caps client.bad osd 'allow rwx pool foo'
 
-
-
-
+* The ``mgr/balancer/max_misplaced`` option has been replaced by a new
+  global option ``target_max_misplaced_ratio`` option that throttles both
```
Comment (xiexingguo) [inline PendingReleaseNotes]: global ``target_max_misplaced_ratio`` option

File: src/mgr/DaemonServer.cc (hunk 1 above)
Comment (liewegas) [reply src/mgr/DaemonServer.cc]: This is effectively what we currently have on 2275

Comment (liewegas) [reply src/mgr/DaemonServer.cc]: moved the check into the next block to improve readability

File: PendingReleaseNotes (hunk 2 above)
Comment (liewegas) [reply PendingReleaseNotes]: fixed, thanks!

File: src/mgr/DaemonServer.cc (hunk 1 above)
Comment (xiexingguo) [reply src/mgr/DaemonServer.cc]: better!

File: src/mgr/DaemonServer.cc [hunk 3]
```diff
@@ -2284,13 +2285,15 @@ void DaemonServer::adjust_pgs()
 			   << " pg_num " << p.get_pg_num()
 			   << " - merge target " << merge_target
 			   << " has upmap" << dendl;
+		  upmaps_to_clear.insert(merge_target);
```
Comment (xiexingguo) [inline src/mgr/DaemonServer.cc]: Ah, I was thinking about the same problem. 
@liewegas Below here I think a better solution is to check and cancel upmap items for merge_target[source] simultaneously.

**Comment by xiexingguo** [comment]
retest this please

File: src/mgr/DaemonServer.cc (hunk 3 above)
Comment (liewegas) [reply src/mgr/DaemonServer.cc]: You mean have the monitor do this based on pg_num_target or something?

Comment (xiexingguo) [reply src/mgr/DaemonServer.cc]: Nope, current implementation looks fine. I mean we could cancel the upmap items in single loop, e.g.:

```c++

		if (osdmap.have_pg_upmaps(merge_target)) {
		  dout(10) << "pool " << i.first
			   << " pg_num_target " << p.get_pg_num_target()
			   << " pg_num " << p.get_pg_num()
			   << " - merge target " << merge_target
			   << " has upmap" << dendl;
                  upmaps_to_clear.insert(merge_target);
		  ok = false;
		}
                if (osdmap.have_pg_upmaps(merge_source)) {
		  dout(10) << "pool " << i.first
			   << " pg_num_target " << p.get_pg_num_target()
			   << " pg_num " << p.get_pg_num()
			   << " - merge source " << merge_source
			   << " has upmap" << dendl;
                  upmaps_to_clear.insert(merge_source);
		  ok = false;
		}

**Comment by neha-ojha** [comment]
retest this please

**Comment by yuriw** [comment]
[wip-yuri2-testing-2018-09-27-2235](https://trello.com/c/WlK4caCc/419-wip-yuri2-testing-2018-09-27-2235)

File: qa/tasks/thrashosds.py [hunk 4]
```diff
@@ -153,6 +155,7 @@ def task(ctx, config):
     config['noscrub_toggle_delay'] = config.get('noscrub_toggle_delay', 2.0)
     # add default value for random_eio
     config['random_eio'] = config.get('random_eio', 0.0)
+    aggro = config.get('aggressive_pg_num_changes', true)
```
Comment (neha-ojha) [inline qa/tasks/thrashosds.py]: `aggro = config.get('aggressive_pg_num_changes', "true")` could be the reason for the teuthology failures.

**Comment by neha-ojha** [comment]
Causing failures such as http://pulpito.ceph.com/yuriw-2018-09-28_12:41:14-rados-wip-yuri2-testing-2018-09-27-2235-distro-basic-smithi/3079733/

**Pull request merged**